cli/diagnostics.rs

```diff
@@ -0,0 +1,641 @@
+
+  /// A hint for fixing the diagnostic.
+  fn hint(&self) -> Option<impl fmt::Display + '_>;
+
+  /// A snippet showing how the diagnostic can be fixed.
+  fn snippet_fixed(&self) -> Option<DiagnosticSnippet<'_>>;
+
+  fn info(&self) -> Cow<'_, [Cow<'_, str>]>;
+
+  /// An optional URL to the documentation for the diagnostic.
+  fn docs_url(&self) -> Option<impl fmt::Display + '_>;
+
+  fn display<'a>(
+    &'a self,
+    sources: &'a dyn SourceTextStore,
+  ) -> DiagnosticDisplay<'a, Self> {
+    DiagnosticDisplay {
+      diagnostic: self,
+      sources,
+    }
+  }
+}
+
+struct RepeatingCharFmt(char, usize);
+impl fmt::Display for RepeatingCharFmt {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    for _ in 0..self.1 {
+      f.write_char(self.0)?;
+    }
+    Ok(())
+  }
+}
+
+/// How many spaces a tab should be displayed as. 2 is the default used for
+/// `deno fmt`, so we'll use that here.
+const TAB_WIDTH: usize = 2;
+
+struct ReplaceTab<'a>(&'a str);
+impl fmt::Display for ReplaceTab<'_> {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    let mut written = 0;
+    for (i, c) in self.0.char_indices() {
+      if c == '\t' {
+        self.0[written..i].fmt(f)?;
+        RepeatingCharFmt(' ', TAB_WIDTH).fmt(f)?;
+        written = i + 1;
+      }
+    }
+    self.0[written..].fmt(f)?;
+    Ok(())
+  }
+}
+
+/// The width of the string as displayed, assuming tabs are 2 spaces wide.
+///
+/// This display width assumes that zero-width-joined characters are the width
+/// of their consituent characters. This means that "Person: Red Hair" (which is
+/// represented as "Person" + "ZWJ" + "Red Hair") will have a width of 4.
+///
+/// Whether this is correct is unfortunately dependent on the font / terminal
+/// being used. Here is a list of what terminals consider the length of
+/// "Person: Red Hair" to be:
+///
+/// | Terminal         | Rendered Width |
+/// | ---------------- | -------------- |
+/// | Windows Terminal | 5 chars        |
+/// | iTerm (macOS)    | 2 chars        |
+/// | Terminal (macOS) | 2 chars        |
+/// | VS Code terminal | 4 chars        |
+/// | GNOME Terminal   | 4 chars        |
+///
+/// If we really wanted to, we could try and detect the terminal being used and
+/// adjust the width accordingly. However, this is probably not worth the
+/// effort.
+fn display_width(str: &str) -> usize {
+  str.width_cjk() + (str.chars().filter(|c| *c == '\t').count() * TAB_WIDTH)
+}
+
+pub struct DiagnosticDisplay<'a, T: Diagnostic + ?Sized> {
+  diagnostic: &'a T,
+  sources: &'a dyn SourceTextStore,
+}
+
+impl<T: Diagnostic + ?Sized> Display for DiagnosticDisplay<'_, T> {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    print_diagnostic(f, self.sources, self.diagnostic)
+  }
+}
+
+// error[missing-return-type]: missing explicit return type on public function
+//   at /mnt/artemis/Projects/github.com/denoland/deno/test.ts:1:16
+//    |
+//  1 | export function test() {
```

Comment:
Do you think in the future we could syntax highlight this code?

Reply:
Absolutely!